driver/driver.go

```diff
@@ -4,69 +4,92 @@
 package driver
 
 import (
+	"errors"
 	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"strings"
+	"time"
 
+	"context"
+
+	"github.com/Microsoft/go-winio"
 	"github.com/Microsoft/hcsshim"
 	log "github.com/Sirupsen/logrus"
 	"github.com/codilime/contrail-windows-docker/common"
 	"github.com/codilime/contrail-windows-docker/controller"
 	"github.com/codilime/contrail-windows-docker/hns"
+	"github.com/codilime/contrail-windows-docker/hnsManager"
+	dockerClient "github.com/docker/docker/client"
 	"github.com/docker/go-plugins-helpers/network"
-	"github.com/docker/go-plugins-helpers/sdk"
 )
 
 type ContrailDriver struct {
-	controller *controller.Controller
-	HnsID      string
+	controller     *controller.Controller
+	hnsMgr         *hnsManager.HNSManager
+	networkAdapter string
+	listener       net.Listener
 }
 
-func NewDriver(subnet, gateway, adapter string, c *controller.Controller) (*ContrailDriver,
-	error) {
+func NewDriver(adapter string, c *controller.Controller) (*ContrailDriver, error) {
 
-	subnets := []hcsshim.Subnet{
-		{
-			AddressPrefix:  subnet,
-			GatewayAddress: gateway,
-		},
+	d := &ContrailDriver{
+		controller:     c,
+		hnsMgr:         &hnsManager.HNSManager{},
+		networkAdapter: adapter,
 	}
+	return d, nil
+}
+
+func (d *ContrailDriver) StartServing() error {
 
-	configuration := &hcsshim.HNSNetwork{
-		Name:               common.NetworkHNSname,
-		Type:               "transparent",
-		Subnets:            subnets,
-		NetworkAdapterName: adapter,
+	pipeConfig := winio.PipeConfig{
+		// This will set permissions for Service, System, Adminstrator group and account to
+		// have full access
+		SecurityDescriptor: "D:(A;ID;FA;;;SY)(A;ID;FA;;;BA)(A;ID;FA;;;LA)(A;ID;FA;;;LS)",
+		MessageMode:        true,
+		InputBufferSize:    4096,
+		OutputBufferSize:   4096,
 	}
 
-	hnsID, err := hns.CreateHNSNetwork(configuration)
-	if err != nil {
-		return nil, err
+	var err error
+	pipeAddr := "//./pipe/" + common.DriverName
+	if d.listener, err = winio.ListenPipe(pipeAddr, &pipeConfig); err != nil {
+		return err
 	}
 
-	d := &ContrailDriver{
-		controller: c,
-		HnsID:      hnsID,
+	if err := os.MkdirAll(common.PluginSpecDir(), 0755); err != nil {
+		return err
+	}
+
+	url := "npipe://" + d.listener.Addr().String()
+	if err := ioutil.WriteFile(common.PluginSpecFilePath(), []byte(url), 0644); err != nil {
+		return err
 	}
-	return d, nil
-}
 
-func (d *ContrailDriver) Serve() error {
 	h := network.NewHandler(d)
+	go h.Serve(d.listener)
 
-	config := sdk.WindowsPipeConfig{
-		// This will set permissions for Service, System, Adminstrator group and account to have full access
-		SecurityDescriptor: "D:(A;ID;FA;;;SY)(A;ID;FA;;;BA)(A;ID;FA;;;LA)(A;ID;FA;;;LS)",
+	// wait for listener goroutine to spin up. I don't see more elegant way to do this.
+	time.Sleep(time.Second * 1)
```

Comment:
How would that help? Still, you would need to make sure that variable is changed atomically with listener starting to serve.